distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso

```diff
@@ -484,16 +487,20 @@ type Enso_File
     list : Text -> Boolean -> Vector Enso_File
     list self (name_filter:Text="") recursive:Boolean=False =
         if self.is_data_link then Data_Link_Helpers.interpret_data_link_target_as_file self . list name_filter=name_filter recursive=recursive else
-            if name_filter != "" then Error.throw (Illegal_Argument.Error "Filtering by name is currently not supported in Enso_File.") else
-                if recursive then Error.throw (Illegal_Argument.Error "Recursive listing is currently not supported in Enso_File.") else
-                    if self.is_directory.not then Error.throw (Illegal_Argument.Error "Cannot `list` a non-directory.") else
-                        # Remove secrets from the list - they are handled separately in `Enso_Secret.list`.
-                        assets = list_assets self . filter f-> f.asset_type != Enso_Asset_Type.Secret
-                        results = assets.map asset->
-                            file = Enso_File.Value (self.enso_path.resolve asset.name)
-                            Asset_Cache.update file asset
-                            file
-                        results.sort on=.path
+            if recursive then Error.throw (Illegal_Argument.Error "Recursive listing is currently not supported in Enso_File.") else
+                if self.is_directory.not then Error.throw (Illegal_Argument.Error "Cannot `list` a non-directory.") else
+                    # Remove secrets from the list - they are handled separately in `Enso_Secret.list`.
+                    assets = list_assets self . filter f-> f.asset_type != Enso_Asset_Type.Secret
+                    filtered_assets = if name_filter == "" then assets else
+                        used_filter = if recursive.not || name_filter.contains "**" then name_filter else
+                            (if name_filter.starts_with "*" then "*" else "{**/,}") + name_filter
+                        matcher = File_Utils.matchPath "glob:"+used_filter
+                        assets.filter a-> matcher.matches (Path.of (a.name.drop 6))
```

Comment:
Would prefer avoiding the magic number for how the constant was derived.

perhaps?
```suggestion
                        assets.filter a-> matcher.matches (Path.of (a.name.drop Enso_Path.protocol_prefix.length))
```

Reply:
Still I'm not convinced if using Java `Path` for Enso paths is safe - the Enso paths have a different set of limitations regarding used characters than e.g. Windows paths.

Reply:
As far as I can see, currently there seem to be no limitations to usernames.

I just managed to change my cloud username to `radeusgd@a:b` and thus `Enso_File.home` yields `enso://Users/radeusgd@a:b` for me.
![image](https://github.com/user-attachments/assets/6ce58d64-d8aa-4222-884d-f3f076af1923)

Such a Path will not be happily accepted by Java Path, leading to a crash:
```
jshell> Path.of("Users/radeusgd@a:b")
|  Exception java.nio.file.InvalidPathException: Illegal char <:> at index 16: Users/radeusgd@a:b
```

Reply:
I think a reasonable approach for now - the global pattern has just turned into a regex. If we find we need to write our own we can but for now will do.

Reply:
Can we then just do Regex based filtering and avoid using Java `Path`?

As I noted above, specific usernames will lead to this code crashing on user files which I don't think is right. We should either make the code handle such cases or disallow such situations by making naming rules on cloud side stricter (we'd need to add requirements for allowed characters in user name and directory/asset titles - currently these two seem to have almost no checks).

![image](https://github.com/user-attachments/assets/084a7e79-1a97-4c05-924f-9144d7e9b383)

I can easily create files with paths like `enso://radeusgd@a:b/a:b:c$$$()d` which will clearly crash the `Path` based matcher with a raw Java exception.

Reply:
will do as a follow up ticket then